design/007-external-data-handling/007-external-data-handling.md

```diff
@@ -0,0 +1,561 @@
+   project can also easily stage it when needed. As above, the data should _not_
+   be pushed to git-LFS but instead kept on the remote resource.
+
+In all three of these cases, at least two of the access modes described in the
+previous section should be accommodated, namely "copy" and "mount". In the case
+of datasets, it's not clear how "reference" is to be used; however, it doesn't
+seem contradictory to organize data references in a dataset and make them
+accessible via Renku's `Dataset` API.
+
+Note: in the commands below `s3://` is used as an example but all of the support
+libraries referred to above support common APIs for (nearly) all likely storage
+backends and automatically switch between them depending on the protocol of the
+URL. That's a long way of saying that one can replace `s3://` with `hdfs://`,
+`nfs://` or `gcs://` etc.
+
+#### Creating a Renku dataset backed by external storage
+
+```
+renku dataset create s3-backed-dataset --storage s3://server/path/to/bucket
+```
+
+This command creates a dataset and sets its backing storage to be S3 instead of
+the default git-LFS. Locally, the directory `data/s3-backed-dataset` is created
+and added to `.gitignore`.
+
+```
+renku dataset add s3-backed-dataset /path/to/file/on/filesystem
+```
+
+This adds a file from the local filesystem to the dataset. It is _copied_ to
+`data/s3-backed-dataset`. (Should it also be pushed to the remote storage
+immediately?)
+
+#### Adding external data to a Renku Dataset
+
+This situation is different than the previous one in that remote data already
+exists but the user wants to bring it into a Renku context. The usual `renku
+dataset add` is used but Renku interprets the `s3://` URL and automatically
+registers it as an external dataset:
+
+```
+renku dataset add --create external-dataset s3://server/path/to/bucket
+-- prompt --
+Access Key: 1234
+Secret Key: abcd
+Store credentials? (y/N): y
+-- output --
+Credentials stored in ~/.renku/renku.ini
+```
+
+On the filesystem, the command creates `data/external-dataset` as is normal for Renku
+datasets. Importantly, `data/external-dataset` is also added to `.gitignore` to avoid
+exploding the git repository. 💥
+
+Note that by adding the data, only the metadata objects are created, the data is
+not actually staged yet.
+
+#### Getting information about the dataset:
+
+```
+renku dataset show external-dataset
+-- output --
+Name: external-dataset
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: external-dataset
+Description: Data that is very large. Indeed.
+Data Sources:
+    - type: s3
+      url: s3://server/path/to/bucket
+      location: None
+```
+
+Note that `location: None` because the dataset has not been staged yet. See
+below for details on that.
+
+#### Staging external data:
+
+External data has to somehow be brought into the system where Renku is running.
+This seems to be a "staging" operation, but we could also call it something
```

Comment:
What about "load"?

Reply:
then it would be, e.g. `renku storage load --mount`? Seems like "load" and "mount" mean two different things.

Reply:
I'm  not sure what you mean- is it bad that "load" and "mount" mean two different things? Loading via copying, vs loading via mounting.

Reply:
yes to me it seems that you would either load or mount the data, in which case `load --mount` doesn't make sense. But maybe `load` is less confusing than `stage`.

Reply:
Mmm yes, I see what you mean. `stage` makes me think of git, and there, `stage` means preparing to send _out_ (`commit` and `push`), whereas in this context we mean more of a pull _in_. ... perhaps we could use `pull`?

Reply:
@lokijuhy I suggest we leave this as is for now since we can't seem to find (agree on) a better alternative. I suspect that the true meaning/naming will crystalize once we actually try to use such a functionality.